src/mds/MDSContext.h

```diff
@@ -135,11 +137,36 @@ class C_IO_Wrapper : public MDSIOContext
   C_IO_Wrapper(MDSRank *mds_, MDSInternalContextBase *wrapped_) : MDSIOContext(mds_), wrapped(wrapped_) {
     assert(wrapped != NULL);
   }
+
+  ~C_IO_Wrapper() {
+    delete wrapped;
+  }
+
   virtual void finish(int r) {
     wrapped->complete(r);
   }
 };
 
+/**
+ * Like C_OnFinisher but cancellable, in which case we do
+ * cleanup manually instead of handling it to the finisher.
+ */
+class MDSIOFinisher: public Context {
+  C_IO_Wrapper *con;
+  Finisher *fin;
+public:
+  MDSIOFinisher(C_IO_Wrapper *c, Finisher *f) : con(c), fin(f) {
+    assert(fin != NULL);
+    assert(con != NULL);
+  }
+  void finish(int r) {
+    if (r == -ECANCELED) {
```

Comment:
Does MDSIOFinisher still need to exist at all?  Couldn't you just do deletion of C_IO_Wrapper now?